calicoctl/commands/printer.go

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2016 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package commands
+
+import (
+	"fmt"
+	"io/ioutil"
+	"reflect"
+	"strings"
+
+	"bytes"
+	"encoding/json"
+	"os"
+	"text/tabwriter"
+	"text/template"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	"github.com/tigera/libcalico-go/calicoctl/resourcemgr"
+	"github.com/tigera/libcalico-go/lib/api/unversioned"
+)
+
+type resourcePrinter interface {
+	print(resources []unversioned.Resource) error
+}
+
+// JSON output formatter
+type resourcePrinterJSON struct{}
+
+func (r resourcePrinterJSON) print(resources []unversioned.Resource) error {
+	// For simplicity convert the slice of resources (which may contain resource lists)
+	// to expand any lists - so we just work with a slice of real resource types.
+	resources = convertToSliceOfResources(resources)
+	if output, err := json.MarshalIndent(resources, "", "  "); err != nil {
+		return err
+	} else {
+		fmt.Printf("%s\n", string(output))
+	}
+	return nil
+}
+
+//YAML output formatter
+type resourcePrinterYAML struct{}
+
+func (r resourcePrinterYAML) print(resources []unversioned.Resource) error {
+	// For simplicity convert the slice of resources (which may contain resource lists)
+	// to expand any lists - so we just work with a slice of real resource types.
+	resources = convertToSliceOfResources(resources)
+	if output, err := yaml.Marshal(resources); err != nil {
+		return err
+	} else {
+		fmt.Printf("%s", string(output))
+	}
+	return nil
+}
+
+// Table output formatter
+type resourcePrinterTable struct {
+	// The headings to display in the table.  If this is nil, the default headings for the
+	// resource are used instead (in which case the `wide` boolean below is used to specify
+	// whether wide or narrow format is required.
+	headings []string
+
+	// Wide format.  When headings have not been explicitly specified, this is used to
+	// determine whether to the resource-specific default wide or narrow headings.
+	wide bool
+}
+
+func (r resourcePrinterTable) print(resources []unversioned.Resource) error {
+	glog.V(2).Infof("Output in table format (wide=%v)", r.wide)
+	for idx, resource := range resources {
+		// Get the resource manager for the resource type.
+		rm := resourcemgr.GetResourceManager(resource)
+
+		// If there are multiple resources then make sure we leave a gap
+		// between each table.
+		if idx > 0 {
```

Comment:
Don't we want to do this all the time?  Just do it _after_ the resource is printed.

This way we get a nice gap between the printed table and the next shell prompt, even when only a single resource is being printed.